discontinuous_spline_extraction option

roystgnr opened the pull request:
## Reason
This lets us read IsoGeometric Analysis meshes where the extracted `RATIONAL_BERNSTEIN` elements don't form a conforming C0 mesh, via the trick of not bothering to try to construct a conforming C0 mesh.

Refs #18768

## Design
MOOSE input file interface to the recent (I meant to get this committed right after the last submodule update...) libMesh API.

## Impact
No change for users who don't specifically set this option.

roystgnr committed:
discontinuous_spline_extraction file option

This gets us a lot more flexibility with what IGA meshes we can support.

Refs #18768

roystgnr committed:
Discontinuous spline extraction regression test

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1465636/) on c826efe wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24081/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1465650/) on c826efe wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>66a35f</th>
      <th colspan=4>#24081 c826ef</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.08%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>86274</td>
      <td>86280</td>
      <td>+6</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15138</td>
      <td>15136</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24081/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24081/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24081/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on test/tests/meshgenerators/file_mesh_generator/2d_discontinuous_iga.i:
```diff
@@ -0,0 +1,53 @@
+    file = PressurizedCyl_Patch6_4Elem.e
+    discontinuous_spline_extraction = true
+  []
+  allow_renumbering = false
+  parallel_type = replicated
```
should all these limitations be baked in the code somewhere? 
I see max_parallel of 1 as well in the test

roystgnr replied:
The allow_renumbering=false is a limitation of XMLDiff, not of our code here.  But right now we're using exodiff thanks to *other* XMLDiff issues, so I'll reenable renumbering.

I've fixed the parallel and distributed-parallel issues now.

GiudGiud commented on test/tests/meshgenerators/file_mesh_generator/gold/2d_discontinuous_iga_out.e:
how do I see block 1 in paraview?

roystgnr replied:
Try setting the display type to "Point Gaussian".  These are individual NodeElem "elements" hosting discrete spline nodes.

GiudGiud replied:
no luck even making them bigger

roystgnr replied:
Huh.  Worked for me, with Paraview 5.10.1.

You did hide block 0?  In our current Exodus output we're basically decimating the mesh, outputting Lagrange elements instead of Rational-Bernstein-Bezier (and then Paraview decimates it further, downgrading quadratics to linears...), and for this particular mesh the 16 spline nodes end up overlapping 16 of the 25 assembly element nodes.

GiudGiud commented on framework/src/meshgenerators/FileMeshGenerator.C:
```diff
@@ -87,6 +95,9 @@ FileMeshGenerator::generate()
       if (mesh->processor_id() == 0)
       {
+        if (getParam<bool>("discontinuous_spline_extraction"))
+          exreader->set_discontinuous_bex(true);
```
what do we get if the user does not set this parameter properly on a mesh that has these features?

roystgnr replied:
It depends.  ;-)

If we have an input mesh for which the assembly elements form a conforming mesh, and all our IGA constraints are implicit in the IGA input mesh, then we get the same FE space, but with fewer constraint equations from the IGA level, because the C0-continuity level of that constraint is maintained by elements simply sharing nodes rather than by redundant equations for each "copy" of a common node.  Global solves on that FE space give the same results to within FP error; any local operations projected back to the spline nodes only give the same results to within FE approximation error (the pseudoinverse involved in that projection minimizes an algebraic norm rather than a nice basis-independent Sobolev norm).

If we have an input mesh for which the assembly elements get IGA constraints from an explicit "extraction operator" (not in this PR, but coming in the next), then when we attempt to load the extraction operator you get a nice message about the numbering not matching.

If we have an input mesh for which the assembly elements form a conforming mesh, and all our IGA constraints are implicit in the IGA input mesh ... I don't actually recall what happens.  I had to put this down for too long to work on other things, and now I can't find a mesh fitting that case.  IIRC you get a somewhat confusing error message at the libMesh level.

I've asked Coreform for a new mesh like that, and if they don't have anything handy I'll try cooking one up myself to see if the failure is user-friendly enough or fix it if not.  In the "not" case, any changes will have to be at the libMesh level, not here.

roystgnr committed:
Mark *why* we turn off parallelism and renumbering

roystgnr committed:
Rename test

I like being able to use `--re _iga` with `run_tests`

roystgnr committed:
Add some exodiff IGA tests

The discontinuous IGA option is giving me problems in parallel, and we
don't have enough test coverage (at the MOOSE level, vs the libMesh
level) of parallel IGA, so let's fix that.

These ExodusII outputs are kind of garbage, since they're turning
properly weighted Rational-Bernstein-Bezier elements as plain Lagrange
elements, but they're good enough to tell whether something is changing
in a regression test.

roystgnr commented:
If a comment doesn't unstale this I'll force push my current version soon.  I haven't actually forgotten it; I just hadn't found time to dig in to the issues I'm seeing until last week.

roystgnr committed:
Add a 1D discontinuous IGA test

This seems to be enough to reproduce the problems we're having in
parallel, and it should be much easier to debug.

roystgnr committed:
Add a discontinuous IGA test with exact solution

Of course *this* one works fine in parallel.  But maybe it'll be useful
for debugging any future regressions more quickly.

roystgnr committed:
Add a discontinuous-IGA case with an L2 problem

*This* one works in parallel!

roystgnr committed:
Put test_2edge.e coords in *projective* space

This is what I get for trying to write IGA files by hand.

roystgnr committed:
Better IGA tests for exposing bugs

roystgnr committed:
Updated gold outputs for Exodus IGA test changes

roystgnr committed:
Re-gold IGA tests affected by libMesh caching bug

roystgnr committed:
IGA on 32-bit ids is fine with the new libMesh

roystgnr committed:
IGA tests can all be enabled with the new libMesh

roystgnr committed:
Update VTK golds for new-libMesh compatibility

We've added a libmesh_node_ids array so we can preserve those in future
reads, but that's enough to confuse the XML differ.

roystgnr committed:
Just allow_renumbering on new tests

If we switch these back to XMLDiff then we'll have to cope with it's
limitations, but that can wait until then.

roystgnr committed:
Make SQA requirements unique
